### ISSUE TITLE
Provide a way to access client information (like address) from an RPC handler

### ISSUE THREAD

**Issue by albertito** [grpc/grpc-go#334]
Having a way to access the client information (like the remote address, client SSL certificate, etc.) from the RPC handlers can be very useful.

For example, this can be used to implement things like logging, tracing, per-client rate limiting and customized ACLs.

Maybe a function to get the credentials from the context, and making sure we can get from them to the relevant info (e.g. have a path from TransportAuthenticator to its corresponding TLSInfo, and I suppose the analogous for oauth would be to get the token)?

**Comment by iamqizhao** [comment]
If you are using a TLS-secured transport, you can get the auth info from context on server side by using https://github.com/grpc/grpc-go/blob/master/credentials/credentials.go#L98.

In general, we will have a way to access peer info and connection info in near future.

**Comment by zj8487** [comment]
+1 for ```access peer info and connection info in near future ```

**Comment by sigmonsays** [comment]
+1

**Comment by mwitkow** [comment]
Is this something that could be slated for GA? We'd be happy to contribute to make this happen.

**Comment by mwitkow** [comment]
For our own internal purposes we managed to work around this problem in the following way:

We implemented a custom `credentials.AuthInfo`:
```go
type CallAuthInfo struct {
	RemoteAddr net.Addr
	Parent credentials.AuthInfo
}

func (t CallAuthInfo) AuthType() string {
	return "callinfo"
}
```

And a custom `credentials.TransportAuthenticator` called `CallInfoAuthenticator`, which among other things does:
```go 
func (c *callInfoAuthenticator) ServerHandshake(rawConn net.Conn) (net.Conn, credentials.AuthInfo, error) {
    callInfo := CallAuthInfo{RemoteAddr: rawConn.RemoteAddr()}
    if c.tlsCreds != nil {
        conn, info, err := tlsCreds.ServerHandshake(rawConn)
        if err != nil {
            return nil, nil, err
        }
        callInfo.Parent = info
        return conn, callInfo, nil
    }
    return rawConn, callInfo, nil
}

```

**Comment by iamqizhao** [comment]
fixed by #485
